merkle/smt/hstar3.go

```diff
@@ -0,0 +1,183 @@
+}
+
+// GetReadSet returns the set of all the node IDs that HStar3 requires to read
+// in order to apply the passed-in updates and propagate node hash changes up
+// to the root(top) node(s) at the specified depth. The updates must have been
+// processed by Prepare.
+//
+// Note: The returned map could have bool value type, but []byte allows the
+// caller to reuse this map for filling in the hashes for the Update method.
+//
+// TODO(pavelkalinnikov): Return only tile IDs.
+func (h HStar3) GetReadSet(updates []NodeUpdate, top uint) map[tree.NodeID2][]byte {
+	ids := make(map[tree.NodeID2][]byte)
+	// For each node, add all its ancestors' siblings, down to the given depth.
+	for _, upd := range updates {
+		for id, d := upd.ID, h.depth; d > top; d-- {
+			pref := id.Prefix(d)
+			if _, ok := ids[pref]; ok {
+				// Delete the prefix node because its original hash does not contribute
+				// to the updates, so should not be read.
+				delete(ids, pref)
+				// All the upper siblings have been added already, so skip them.
+				break
+			}
+			sib := pref.Sibling()
+			ids[sib] = nil
+		}
+	}
+	return ids
+}
+
+// Update applies the given updates to a sparse Merkle tree. Requires a
+// previous Prepare invocation on the same updates slice. Returns an error if
+// any of the NodeStorage.Get calls does so, for example if a node is missing.
+//
+// Returns the slice of updates at the top level of the sparse Merkle tree
+// induced by the provided lower level updates. Typically it will contain only
+// one item for the root hash of a tile or a (sub)tree, but the caller may
+// arrange multiple subtrees in one call, in which case the corresponding
+// returned top-level updates will be sorted lexicographically by node ID.
+//
+// Note that Update invocations can be chained. For example, a bunch of HStar3
+// instances at depth 256 can return updates for depth 8 (in parallel), which
+// are then merged together and passed into another HStar3 at depth 8 which
+// computes the overall tree root update.
+//
+// For that reason, Update doesn't invoke NodeStorage.Set for the topmost level
+// nodes. If it did then chained Updates would Set the borderline nodes twice.
+//
+// Warning: This call modifies the updates slice in-place, so the caller must
+// ensure to not reuse it.
+func (h HStar3) Update(updates []NodeUpdate, top uint, ns NodeStorage) ([]NodeUpdate, error) {
+	for d := h.depth; d > top; d-- {
+		var err error
+		if updates, err = h.updateAt(updates, d, ns); err != nil {
+			return nil, fmt.Errorf("depth %d: %v", d, err)
+		}
+	}
+	return updates, nil
+}
+
+// updateAt applies the given node updates at the specified tree level.
+// Returns the updates that propagated to the level above.
+func (h HStar3) updateAt(updates []NodeUpdate, depth uint, ns NodeStorage) ([]NodeUpdate, error) {
+	// Apply the updates.
+	for _, upd := range updates {
+		ns.Set(upd.ID, upd.Hash)
+	}
+	// Calculate the updates that propagate to one level above.
+	newLen := 0
```

Comment:
I think a comment on newLen that it will be in the range [length(updates)/2, length(updates)] & it shrinks whenever two updates have the same parent.